Feature indent depth check

yubaoquan opened the pull request:
#### PR checklist

- [x] Addresses an existing issue: #2814
- [x] Enhancement
  - [x] Includes tests
- [x] Documentation update

#### Overview of change:
Add ability to check the indent depth. 

For example:
```
function foo() {
    var a = 1 // 1 indent depth
    var b = function() {
        var c = 2 // 2 indent depth
    } // 1 indent depth
} // 0 indent depth
```

#### Is there anything you'd like reviewers to focus on?

Maybe not all use cases in the test files are covered.

#### CHANGELOG.md entry:

<!-- optional (example: "[new-rule] `arrow-return-shorthand`") -->
<!-- suggested tags: [new-rule], [new-rule-option], [new-fixer], [bugfix], [enhancement], [api], [rule-change], [no-log] -->

[new-fixer]
[enhancement]

ajafff commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +128,161 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+}
+
+function recursiveGetNodeAt(root: ts.Node, pos: number): ts.Node | undefined {
+    let result: ts.Node | undefined = root.getChildAt(pos);
```
this does not what you intended. it gets the `n`th child, not the child at position `n`

yubaoquan replied:
It just works. Not found a test case can prove this usage is wrong.

ajafff commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +128,161 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+            if (node.getFullStart() < pos && node.getEnd() >= pos) {
+                result = recursiveGetNodeAt(node, pos);
+            }
+            ts.forEachChild(node, cb);
```
you won't find the node if you recurse here, because the parent is already out of the range

yubaoquan replied:
ok, code removed

ajafff commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +128,161 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+    const currentNodePos: number = pos + indentEnd;
+    let currentNode: ts.Node | undefined = recursiveGetNodeAt(sourceFile, currentNodePos);
+    let depth = 0;
+    if (currentNode !== undefined && inTsx(sourceFile)) {
```
Is there a reason to do nothing at all in .tsx files? And should the same apply to .jsx?
If there is a good reason, you could consider moving this check to the top of `walk` to avoid unnecessary work.

yubaoquan replied:
Added the check for jsx syntax.

ajafff commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +128,161 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+    indentEnd: number;
+    line: string;
+    sourceFile: ts.SourceFile;
+}): boolean {
```
consider moving this function inside `walk` so it closes over the variables like `ctx`, `tabs`, `sourceFile` and so on.

yubaoquan replied:
done

ajafff commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +128,161 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+        ts.SyntaxKind.TemplateHead,
+        ts.SyntaxKind.TemplateMiddle,
+        ts.SyntaxKind.TemplateTail,
+    ].indexOf(node.kind) > -1;
```
```ts
node.kind >= ts.SyntaxKind.FirstTemplateToken && node.kind <= ts.SyntaxKind.LastTemplateToken
```

yubaoquan replied:
As you wish

ajafff commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +128,161 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+    let result = 0;
+    let parent: ts.Node | undefined = node.parent;
+    const blockTypes: ts.SyntaxKind[] = [
+        ts.SyntaxKind.Block, // 207
```
these numbers are subject to change with future versions of typescript. you can remove them, because they do not add any value

yubaoquan replied:
removed

ajafff commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +128,161 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+        ts.SyntaxKind.CaseBlock, // 235
+        ts.SyntaxKind.EnumMember, // 264
+        ts.SyntaxKind.CaseClause, // 257
+        // ts.SyntaxKind.IfStatement, // Have no idea on how to check whether a node is in if paren
```
you probably also need `ForStatement`, `ForOfStatement`, `ForInStatement`, `WhileStatement`, `DoStatement` and `WithStatement` but only if they do not contain a block

yubaoquan replied:
Added another function to check text in parens of these statements

ajafff commented on docs/_data/formatters.json:
```diff
@@ -0,0 +1,83 @@
+[
```
if you merge latest master these generated files should be gone

yubaoquan replied:
done

adidahiya commented:
Note that there is already another open PR that tackles the same issue: #3300

yubaoquan commented:
How's the progress?

yubaoquan commented:
Any problems with this pr? @ajafff

ajafff commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +159,130 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+            result++;
+        }
+        if (inStatementParen(parent)) {
+            result ++;
```
some special cases to consider:

```ts
// multiline condition, works well for `if`, but probably doesn't for `while`
while(foo() &&
      bar() &&
      baz()) {
    doStuff;
}
// might not be worth fixing, because it can be reformatted to
while(
    foo() &&
    bar() &&
    baz()
) {
    doStuff;
}

// statement continuation on the next line
let foo = bar
    ? baz
    : bas;
foo = "some"
    + "really"
    + "long"
    + "string";

// multiline property access
foo
    .bar
    .baz();
```

yubaoquan replied:
OK, add these into test cases. Test passed.

ajafff commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +159,130 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+    if (
+        node.parent !== undefined
+        && node.parent.kind === ts.SyntaxKind.CallExpression
+        && /^\s*\./.test(line)
```
please add a comment explaining what this condition does. it took me a while to figure it out

yubaoquan replied:
done

ajafff commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +159,130 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+    );
+}
+
+function nodeAtOutside(node: ts.Node): boolean {
```
please add a comment to describe what this function does

yubaoquan replied:
done

ajafff commented:
@yubaoquan sorry for the delay, I got distracted. Unfortunately writing new code is a lot more fun than reviewing someone else's code.

yubaoquan commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +159,130 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+        if (node.getFullStart() < pos && node.getEnd() >= pos) {
+            result = recursiveGetNodeAt(node, pos);
+        }
+        ts.forEachChild(node, cb);
```
This is used to find the closing bracket line. Should not be removed.

yubaoquan commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +159,130 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+    const result: ts.Node | undefined = root.getChildAt(pos);
+    if (result === undefined) {
+        ts.forEachChild(root, function cb(node: ts.Node): void {
+            ts.forEachChild(node, cb);
```
Removed.

yubaoquan commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +159,130 @@ function walk(ctx: Lint.WalkContext<Options>): void {
 }
 
+function recursiveGetNodeAt(root: ts.Node, pos: number): ts.Node | undefined {
+    const result: ts.Node | undefined = root.getChildAt(pos);
```
Updated. The `recursiveGetNodeAt` and `recursiveGetNodeAt` are not suitable names with the logic, and the inner logic is not clearly, so I refactor theses two function to one single function `recursiveGetNodeAt`.
The function just find the node of current line. If the pos is between the getFullStart and getEnd, that means the line is the node or is child of the node.

yubaoquan commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +159,130 @@ function walk(ctx: Lint.WalkContext<Options>): void {
     }
 }
 
+function recursiveGetNodeAt(root: ts.Node, pos: number): ts.Node | undefined {
```
Tried, the replacement produce wrong results. The `getTokenAtPosition` will always return a token, which is not expected. In `recursiveGetNodeAt`, the `getChildAt` will return undefined on some positions, and then the function should return the root node. If replace to getTokenAtPosition, the returned value will be unexpected. Anyway I have refactored this function.

yubaoquan commented on src/rules/indentRule.ts:
```diff
@@ -100,12 +100,50 @@ interface Options {
+                const passIndentDepthCheck = inStringTemplate(currentNode) || (line.trim().charAt(0) === "*");
+
+                if (!passIndentDepthCheck && expectedIndentation !== actualIndentation) {
+                    failure = Rule.FAILURE_STRING(tabs ? "tab" : size === undefined ? "space" : `${depth * size} space`);
```
updated

yubaoquan commented on src/rules/indentRule.ts:
```diff
@@ -100,12 +100,50 @@ interface Options {
+                const expectedIndentation: string = tabs ? "\t".repeat(depth)
+                : " ".repeat(size * depth);
+                const actualIndentation = line.slice(0, indentEnd);
+                const passIndentDepthCheck = inStringTemplate(currentNode) || (line.trim().charAt(0) === "*");
```
done

yubaoquan commented on src/rules/indentRule.ts:
```diff
@@ -100,12 +100,50 @@ interface Options {
+            if (currentNode !== undefined && currentNode.getStart() === currentNodePos) {
+                depth = getNodeDeepth(currentNode, line);
+            } else {
+                currentNode = recursiveGetNodeAtEnd(sourceFile, currentNodePos);
```
In `if` block, the condition matches the opening part of a block, like `module TestModule {`; 
In `else` block, the condition matches the closing part of a block, like a single `}`.

yubaoquan commented on src/rules/indentRule.ts:
```diff
@@ -100,12 +100,50 @@ interface Options {
+            const currentNodePos: number = pos + indentEnd;
+            let currentNode: ts.Node | undefined = recursiveGetNodeAt(sourceFile, currentNodePos);
+            let depth = 0;
+            if (currentNode !== undefined && currentNode.getStart() === currentNodePos) {
```
done

yubaoquan commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +159,130 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+        && /^\s*<\//.test(line);
+}
+
+function inStringTemplate(node: ts.Node | undefined): boolean {
```
done

yubaoquan commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +159,130 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+    if (node === undefined) {
+        return false;
+    }
+    return node.kind === ts.SyntaxKind.NoSubstitutionTemplateLiteral
```
done

yubaoquan commented:
Updated the code as you said. Thank you for your review and advices!

yubaoquan commented:
Something is wrong with CI, how to fix it? Should I merge the master branch into this branch?

> error Couldn't find the binary npm-run-all -p compile:core compile:test -s compile:scripts

ajafff commented:
> Something is wrong with CI, how to fix it? Should I merge the master branch into this branch?

Exactly. This is fixed on master, you just need to merge

yubaoquan commented:
how's it going?

yubaoquan commented:
Do we have an estimate of when this might be merged? @ajafff @adidahiya

ajafff commented:
@yubaoquan before jumping right back to reviewing the code, I tried how it works. There are several false positives that need to be addressed:

multiline arguments:
```ts
foo(
    bar, // 0 space indentation expected
    baz, // 0 space indentation expected
);

new Foo(
    bar, // 0 space indentation expected
    baz, // 0 space indentation expected
);
```

continuation on the next line:
```ts
let someReallyLongVariableName =
    anotherReallyLongVariableName; // 0 space indentation expected

return foo
    && bar // 0 space indentation expected
    && baz; // 0 space indentation expected

someArray.filter((element) =>
    doStuff(element)); // 0 space indentation expected
// or
const someFn = (element) =>
    doStuff(element); // 0 space indentation expected
```

multiline named imports:
```ts
import {
    foo, // 0 space indentation expected
    bar, // 0 space indentation expected
} from "foobar";

// probably the same for 'export'
export {
    foo, // 0 space indentation expected
    bar, // 0 space indentation expected
};
```

else-if:
```ts
if (foo) {
    foo;
} else if (bar) {
    bar; // 8 space indentation expected
} // 4 space indentation expected
```

type literals:
```ts
export interface TestResult {
    directory: string;
    results: {
        [fileName: string]: TestOutput | SkippedTest; // 4 space indentation expected
    };
}
```

ParenthesizedExpression:
```ts
return (
    foo // 0 space indentation expected
);
```

template expressions:
```ts
`foo ${
   foo // 0 space indentation expected
}`
```

template literals:
```ts
foo(
    `foo`, // this line is valid
// next line should have an error, currently it is ignored
  `
bar
`,
// the two lines above are valid, because the indent is inside the template string
);
```

indent wrongly calculated at start of statement:
```ts
// next line starts with one space
 for(;;) {} // 4 space indentation expected
```

Of course everything mentioned above is open for discussion.

yubaoquan commented:
All cases fixed except two below:

```
foo(
    `foo`, // this line is valid
// next line should have an error, currently it is ignored
  `
bar
`,
// the two lines above are valid, because the indent is inside the template string
);
```
1. I don't know what error should we get.
2. I removed indent checking in string template because the indent of the first variable in string template isn't credible (the indent could be part of the string).

ajafff commented:
IMO the opening backtick of a template string should be checked like every other token. We just ignore the indent **inside** the template string, because that could be part of the content.

> What's more, if you got a ~ [1] in a string template, how do you make sure whether it is an error mark or content of the string?

The `*.ts.lint` files are preprocessed to remove all error markers before parsing the file content as TypeScript. That means you can have valid typescript code mistakenly treated as error marker (very unlikely), but not the other way around.

ajafff commented:
Here we go with the next round of false positives:

```ts
        const linter = new Linter({
            fix: false,
            formatter: "prose",
        }); // 12 spaces indentation expected - currently 8
```

still a problem with continuation on the next line:
```ts
        const expectedResult =
            getFailureString(TEST_FILE, 1,  1, "first failure", "first-name") + // 8 space indentation expected - currently 12
            getFailureString(TEST_FILE, 2, 12, "mid failure", "mid-name") + // 8 space indentation expected - currently 12
            getFailureString(TEST_FILE, 9,  2,  "last failure", "last-name"); // 8 space indentation expected - currently 12
```

chained method calls:
```ts
fs.createReadStream("test/files/multiple-fixes-test/multiple-fixes.test.ts")
    .pipe(fs.createWriteStream(tempFile))
    .on("finish", () => {
        doStuff(); // 4 spaces indentation expected
    }); // 0 spaces indentation expected
```

blocks in CaseClause and DefaultClause
```ts
            case ts.SyntaxKind.BinaryExpression: {
                break; // 20 spaces indentation expected - currently 16
            } // 20 spaces indentation expected - currently 12
```

multiline arrow function as argument:
```ts
    return find(sym.declarations, (decl) =>
        isClassLikeDeclaration(decl) ? decl.typeParameters : undefined); // 12 - currenly 8
```

property initializer on the next line:
```ts
class Foo {
    public static FAILURE_STRING_TRAILING =
        "Did not expect single-line type literal to have a trailing ';'."; // 4 spaces indentation expected -
 currently 8
}
```

something with TypeAlias:
```ts
export type Location =
    | ts.PrefixUnaryExpression // 0 spaces indentation expected (for all of them)
    | ts.IfStatement
    | ts.WhileStatement
    | ts.DoStatement
    | ts.ForStatement
    | ts.ConditionalExpression
    | ts.BinaryExpression;
```

something with abstract class:
```ts
// only reproducible if the class declaration is not on the first line
abstract class Foo {} // 4 spaces indentation expected
```

multiline type arguments:
```ts
type T = Map<
    string, // 0 space indentation expected
    string // 0 space indentation expected
>;
```

---

I like how you handle nested conditional expressions:
```ts
return foo
    ? foo
    : bar
        ? bar
        : baz;
```
Though others prefer the following, which is now invalid. I don't have a strong opinion for one of them.
```ts
return foo ? foo
    : bar ? bar
    : baz;
```

---

I actually feel kind of sorry for you, because your first contribution turns out to be that difficult. I really appreciate the effort you put into this PR.

yubaoquan commented:
My logic of calculating the depth of code after chaining call is not good. I need time to refactor that part. Recently pretty busy, so it will take weeks before committing next pr.

yubaoquan commented:
Pretty hard to cover all the code styles. Pretty hard to debug this time. Anyway, fixed all cases you mentioned last time, and added the corresponding test cases.

As for the conditional expressions, I have no idea, should it to be an configurable option in this rule?

yubaoquan commented:
test/rules/trailing-comma/singleline-always/test.ts.lint: Failed! 
It failed on a test which is not of indent check, is it matters?
I have already merged the master branch,  how to fix it now?

ajafff commented:
You cannot fix the failing test, see https://github.com/palantir/tslint/pull/3494#issuecomment-345191452

adidahiya commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +156,383 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+    return result;
+}
+
+// function isSameLineWithConditionalExpr(sourceFile: ts.SourceFile, nodePos: number, parent: ts.Node): boolean {
```
remove commented code

yubaoquan replied:
done

adidahiya commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +156,383 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+foo()
+.then()
+In this case, `.then` should have one more indent.
+Maybe this should depend on a configuration.
```
use proper JSDoc formatting please

adidahiya replied:
also, is this last line a question for code reviewers? I don't think this is a helpful comment without more elaboration. Either remove it or make the "maybe" part more clear.

yubaoquan replied:
done

adidahiya commented on src/rules/indentRule.ts:
```diff
@@ -121,6 +156,383 @@ function walk(ctx: Lint.WalkContext<Options>): void {
+ *      // xxx
+ *  }
+ *  In this case, the `Foo` is regarded as should have the same indent depth with `class`.
+ *  Should it have one more depth? I'm not sure.
```
use proper JSDoc formatting -- there is an extra whitespace after each `*` here

yubaoquan replied:
done

ajafff commented:
@adidahiya 
> I personally prefer to leave such formatting to tools like prettier :)

Is that a -1 on this PR? If so, we can stop here and instead point people to [`ter-indent` from `tslint-eslint-rules`](https://github.com/buzinas/tslint-eslint-rules/blob/master/src/docs/rules/terIndentRule.md)

adidahiya commented:
Sorry, no, it's not an explicit -1 on this PR.

There's a lot of demand for this feature in core TSLint, so it makes sense to add it, but we (project collaborators) might provide a lower amount of support for it than other core features. It might even deserve a callout in the docs that says the fixer for this rule is "experimental" (defer to your judgement here -- I haven't looked very closely at the CR progress and the false positives identified here).

I think in the future we should break up the core rule set into multiple packages, one of which would be formatting rules. Users of TSLint + Prettier would be able to disable the formatting rules all at once and we could set users' expectations of support accordingly. But of course that's out of scope for this PR.

yubaoquan commented:
When can this be merged? Or lets think more test cases? @ajafff

ajafff commented:
Ok, let's continue:

```ts
const expectedResult = 
    `foo`; // 0 spaces expected

return execRunner(
    {
        foo,
    }, // 0 spaces expected
);

const options = {
    foo; // 8 spaces expected
}; // 4 spaces expected

type Options = Record< // 4 spaces expected
    "branch" | "decl" | "operator" | "module" | "separator" | "restSpread" | "type" | "typecast" | "typeOperator" | "preblock",
    boolean>;
type Foo = boolean; // 4 spaces expected

return (a.dotDotDotToken !== undefined) === (b.dotDotDotToken !== undefined) &&
        (a.questionToken !== undefined) === (b.questionToken !== undefined); // 0 spaces expected

const couldBeSwitch = everyCase(node, (expr) => {
    casesSeen++;
    if (switchVariable !== undefined) {
        return nodeEquals(expr, switchVariable, sourceFile);
    } else {
        switchVariable = expr; // 12 spaces expected
        return true;
    }
});

doStuff([
    foo, // 8 spaces expected
]); // 4 spaces expected
```

I think that's it for now.

---

I think we better add a new option for checking the indent level. This new behavior could make a lot users insane when enabled by default.

yubaoquan commented:
Done. New test cases added, bug fixed. 
The cases are growing bigger and bigger.

yubaoquan commented:
I think a better way to check indent is making a tree, each line of the source file corresponds to a leaf on the tree. The depth of the leaf is the indent depth of the line. So the whole thing would be much easier. But I'm not good at making such a tree.

cyberhck commented:
is it almost ready to merge now? any estimation on when this can get merged?

yubaoquan commented:
@ajafff It's time to find new bugs, need your report! thanks :)

erikkemperman commented:
@yubaoquan @ajafff I'd really *love* to see this feature released, so for what it's worth:

- I've fetched this branch and merged current master into it;
- Ran `yarn install --pure-lockfile && yarn run validate` (all tests still pass);
- Used `yarn link` to use this version in my project;
- Verified that it finds various indentation errors I intentionally added;
- Verified that it fixes indentation errors I intentionally added;

Of course there are many variations of nested blocks and statemtent continuations and so on and I can't try everything, but as far as I did try this seems to work well!

giladgray commented:
@yubaoquan please update this branch so we can review it, or close it if no longer relevant. we will close this if we do not hear from you in two weeks.

I am inclined to reject this feature for two reasons:
1. soooo much code is hard to review and trust.
2. [Prettier](https://www.npmjs.com/package/tslint-config-prettier) handles this feature elegantly and has fixer support.

johnwiseheart commented:
Closing due to age and inactivity.
